### PR TITLE
fix: Extra bounds checks on AcpiTableUpdate copy

### PR DESCRIPTION
**Pull request by bejeanmo**
When AcpiTableUpdate copies a caller provided table, check for valid header length and possible source and target buffer overflow.

### CODE REVIEW

**Review by lsubash — APPROVED** [review]

File: BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c [hunk 1]
```diff
@@ -211,20 +211,39 @@ AcpiTableUpdate (
     AcpiHdr = (EFI_ACPI_DESCRIPTION_HEADER *)(AcpiTable + Size);
 
     //
-    // Verify Checksum
+    // Verify the ACPI description header itself fits within the input buffer
```
Comment (tchong7) [inline BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c:214]: Do we add new negative testing test case to ensure it take effect?  Thanks for the good changes to harden the buffer overflow attack with malform header length even when check sum is correct.

**Review by tchong7 — APPROVED** [review]

File: BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c [hunk 2]
```diff
@@ -211,20 +211,39 @@ AcpiTableUpdate (
     AcpiHdr = (EFI_ACPI_DESCRIPTION_HEADER *)(AcpiTable + Size);
```
Comment (spkamat) [inline BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c:211]: This change is not part of this PR. However, a NULL check for AcpiHdr would be needed here?

File: BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c [hunk 3]
```diff
@@ -211,20 +211,39 @@ AcpiTableUpdate (
...
 
     //
-    // Verify Checksum
+    // Verify the ACPI description header itself fits within the input buffer
+    // before reading any header fields (e.g. AcpiHdr->Length).
     //
-    if (CalculateSum8 ((UINT8 *)AcpiHdr, AcpiHdr->Length) != 0) {
+    if (Length - Size < sizeof (EFI_ACPI_DESCRIPTION_HEADER)) {
```
Comment (spkamat) [inline BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c]: nitpick: braces around L-S to improve readability.

File: BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c [hunk 4]
```diff
@@ -211,20 +211,39 @@ AcpiTableUpdate (
...
-      Size +=  AcpiHdr->Length;
-    } else {
+    //
+    // Verify the full table (as declared by the header) fits within the input
+    // buffer.  Also reject a declared Length smaller than the header itself.
+    // Without this check a crafted AcpiHdr->Length causes CalculateSum8 and
+    // CopyMem to read beyond the caller-supplied buffer (OOB read).
+    //
+    if ((AcpiHdr->Length < sizeof (EFI_ACPI_DESCRIPTION_HEADER)) ||
+        (AcpiHdr->Length > Length - Size)) {
```
Comment (spkamat) [inline BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c]: ditto for L-S

File: BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c (hunk 2 above)
Comment (bejeanmo) [reply BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c:211]: AcpiTable is NULL checked so I think it's fine.

File: BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c (hunk 3 above)
Comment (bejeanmo) [reply BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c]: Done

File: BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c (hunk 4 above)
Comment (bejeanmo) [reply BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c]: Done
